File: src/ui/selection.ts
import { union } from '../util/objTransforms';
import { Id, StringSlice, NodeIndicatorInPlot } from '../content/types';
import { calculateNodeCenterOnPlot, PlotRect } from './coords';
import { PlotCoords, PositionedNode, PositionedTree } from '../content/positioned/types';
import { UnpositionedPlot } from '../content/unpositioned/types';

export type NodeSelectionInPlot = { nodeIndicators: NodeIndicatorInPlot[] };
export type SliceSelectionInPlot = { treeId: Id, slice: StringSlice };
export type SelectionInPlot = NodeSelectionInPlot | SliceSelectionInPlot;

export const isNodeSelection = (selection: SelectionInPlot): selection is NodeSelectionInPlot => 'nodeIndicators' in selection;
export const isSliceSelection = (selection: SelectionInPlot): selection is SliceSelectionInPlot => 'slice' in selection;

export type NodeSelectionAction = 'select' | 'adopt' | 'disown';
export type NodeSelectionMode = 'set' | 'add';

export const applySelection = (
  mode: NodeSelectionMode, newNodeIndicators: NodeIndicatorInPlot[], existingNodeIndicators?: NodeIndicatorInPlot[]) =>
  mode === 'add' ? union(existingNodeIndicators || [], newNodeIndicators)
    : newNodeIndicators;

/**
 * Returns whether the node indicated by the given indicator exists in the given plot.
 */
const indicatorTargetExistsInPlot = (plot: UnpositionedPlot) => (nodeIndicator: NodeIndicatorInPlot) =>
  !!plot.trees[nodeIndicator.treeId] && !!plot.trees[nodeIndicator.treeId].nodes[nodeIndicator.nodeId];

/**
 * Returns a copy of the given node selection including only nodes matching the given predicate.
 */
const filterNodesInSelection = (
  selection: NodeSelectionInPlot,
  predicate: (nodeIndicator: NodeIndicatorInPlot) => boolean,
): SelectionInPlot =>
  ({ nodeIndicators: selection.nodeIndicators.filter(predicate) });

/**
 * Removes nonexistent nodes from the given selection, based on the given plot.
 */
export const pruneSelection = (selection: SelectionInPlot, plot: UnpositionedPlot): SelectionInPlot =>
  isSliceSelection(selection)
    ? (plot.trees[selection.treeId] ? selection : { nodeIndicators: [] })
    : filterNodesInSelection(selection, indicatorTargetExistsInPlot(plot));

/**
 * Returns whether the given point is inside the given rectangle.
 */
const isPointInPlotRect = (rect: PlotRect) => (coords: PlotCoords) =>
  coords.plotX >= rect.topLeft.plotX && coords.plotY >= rect.topLeft.plotY &&
  coords.plotX <= rect.bottomRight.plotX && coords.plotY <= rect.bottomRight.plotY;

export const isNodeInRect = (rect: PlotRect) => (tree: PositionedTree) => (node: PositionedNode) =>
  isPointInPlotRect(rect)(calculateNodeCenterOnPlot(tree)(node));

File: npbloom-core/src/commonMain/kotlin/ui/Selection.kt
@file:OptIn(ExperimentalJsExport::class)

package ui

import content.Id
import content.NodeIndicatorInPlot
import content.StringSlice
import content.positioned.PositionedNode
import content.positioned.PositionedTree
import content.unpositioned.UnpositionedPlot
import kotlin.js.ExperimentalJsExport
import kotlin.js.JsExport
import kotlin.js.JsName

@JsExport
sealed interface SelectionInPlot

@JsExport
data class NodeSelectionInPlot internal constructor(val nodeIndicators: Set<NodeIndicatorInPlot> = emptySet()) :
    SelectionInPlot {
    val nodeIndicatorsAsArray = nodeIndicators.toTypedArray()

    companion object {
        fun fromArray(nodeIndicatorArray: Array<NodeIndicatorInPlot>) = NodeSelectionInPlot(nodeIndicatorArray.toSet())
    }
}

@JsExport
data class SliceSelectionInPlot(val treeId: Id, val slice: StringSlice) : SelectionInPlot

@JsExport
enum class NodeSelectionAction { Select, Adopt, Disown }

@JsExport
enum class NodeSelectionMode { SetSelection, AddToSelection }

@JsName("applySelectionByKtSets")
fun applySelection(
    mode: NodeSelectionMode,
    newNodeIndicators: Set<NodeIndicatorInPlot>,
    existingNodeIndicators: Set<NodeIndicatorInPlot> = emptySet(),
): Set<NodeIndicatorInPlot> = when (mode) {
    NodeSelectionMode.AddToSelection -> existingNodeIndicators + newNodeIndicators
    NodeSelectionMode.SetSelection -> newNodeIndicators
}

/**
 * Removes nonexistent nodes from the given selection, based on the given plot.
 */
internal fun pruneSelection(selection: SelectionInPlot, plot: UnpositionedPlot): SelectionInPlot =
    when (selection) {
        is SliceSelectionInPlot -> if (selection.treeId in plot) selection else NodeSelectionInPlot(emptySet())
        is NodeSelectionInPlot -> NodeSelectionInPlot(selection.nodeIndicators.filter { it in plot }.toSet())
    }

@JsExport
fun isNodeInRect(tree: PositionedTree, node: PositionedNode, rect: RectInPlot) =
    calculateNodeCenterOnPlot(tree, node) in rect
